src/pwr.rs

```diff
@@ -0,0 +1,169 @@
+//! The processor supports multiple voltage scaling modes, from VOS3 (lowest
+//! performance, lowest power consumption) to VOS0 (highest performance,
+//! highest power consumption).
+//!
+//! After reset the system is in VOS3. This power configuration module allows
+//! modes VOS0 to VOS3 to be selected (it defaults to VOS0 after calling
+//! `freeze`.
+//!
+//! ```rust
+//!     let dp = pac::Peripherals::take().unwrap();
+//!
+//!     let pwr = dp.PWR.constrain();
+//!     let pwrcfg = pwr.freeze();
+//!
+//!     assert_eq!(pwrcfg.vos(), VoltageScale::Scale0);
+//! ```
+//!
+use crate::stm32::pwr::voscr::VOS;
+use crate::stm32::pwr::vossr::ACTVOSR;
+use crate::stm32::PWR;
+
+/// Extension trait that constrains the `PWR` peripheral
+pub trait PwrExt {
+    fn constrain(self) -> Pwr;
+}
+
+impl PwrExt for PWR {
+    fn constrain(self) -> Pwr {
+        Pwr {
+            rb: self,
+            target_vos: VoltageScale::Scale0,
+        }
+    }
+}
+
+/// Constrained PWR peripheral
+///
+/// Generated by calling `constrain` on the PAC's PWR peripheral.
+pub struct Pwr {
+    pub(crate) rb: PWR,
+    target_vos: VoltageScale,
+}
+
+/// Voltage Scale
+///
+/// Represents the voltage range feeding the CPU core. The maximum core
+/// clock frequency depends on this value.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum VoltageScale {
+    /// VOS 0 range VCORE 1.30V - 1.40V
+    Scale0,
+    /// VOS 1 range VCORE 1.15V - 1.26V
+    Scale1,
+    /// VOS 2 range VCORE 1.05V - 1.15V
+    Scale2,
+    /// VOS 3 range VCORE 0.95V - 1.05V
+    Scale3,
+}
+
+impl From<VoltageScale> for VOS {
+    fn from(value: VoltageScale) -> Self {
+        match value {
+            VoltageScale::Scale3 => VOS::Vos3,
+            VoltageScale::Scale2 => VOS::Vos2,
+            VoltageScale::Scale1 => VOS::Vos1,
+            VoltageScale::Scale0 => VOS::Vos0,
+        }
+    }
+}
+
+impl From<ACTVOSR> for VoltageScale {
+    fn from(value: ACTVOSR) -> Self {
+        match value {
+            ACTVOSR::Vos3 => VoltageScale::Scale3,
+            ACTVOSR::Vos2 => VoltageScale::Scale2,
+            ACTVOSR::Vos1 => VoltageScale::Scale1,
+            ACTVOSR::Vos0 => VoltageScale::Scale0,
+        }
+    }
+}
+
+/// Power Configuration
+///
+/// Generated when the PWR peripheral is frozen.
+/// longer be changed.
+pub struct PowerConfiguration {
+    pub(crate) vos: VoltageScale,
+}
+
+impl PowerConfiguration {
+    /// Gets the `VoltageScale` which was configured by `Pwr::freeze()`.
+    pub fn vos(&self) -> VoltageScale {
+        self.vos
+    }
+}
+
+/// Internal power methods
+impl Pwr {
+    /// Transition between voltage scaling levels
+    fn voltage_scaling_transition(&self, new_scale: VoltageScale) {
+        // ************************************
+        // TODO: Warning! STM32H503 Errata 2.2.13 states that for Rev A, Z this transition should only
+        //       be executed from RAM
+        //       Move this function into RAM!
```

Comment:
This is a pretty impressive errata! I wonder if any Rev A or Z hardware ever made it out into the wild? Assuming that current dev boards have revision Y parts on them, I wouldn't worry about it too much.

Reply:
Yeah, it's a fun one. I don't know whether Rev A or Z was ever released widely. It seems as though the dev boards that ST sells and the processors that I've used since Q3 2023 are Rev Y. I can probably just leave a note here, instead of a TODO, given how unlikely it seems that someone would actually run into this.